File: frontend/src/features/workspace/types/sse.ts
export type SseConnectionStatus =
  | "idle"
  | "connecting"
  | "open"
  | "closed"
  | "error";

/**
 * Persisted SSE connection preferences for a workspace tab.
 */
export type SseConnectionConfig = {
  /**
   * Sent as the `Last-Event-ID` request header on connect/reconnect so the
   * server can resume the stream after the last seen event id.
   */
  lastEventId: string;

  /**
   * When set, only events whose `event:` field matches are shown in the
   * stream log. Empty string means accept all event types.
   */
  eventTypeFilter: string;

  /**
   * Include cookies on the cross-origin request (`credentials: 'include'`).
   */
  withCredentials: boolean;

  /**
   * When the connection drops unexpectedly, attempt to reconnect using the
   * latest received event id as `Last-Event-ID`.
   */
  autoReconnect: boolean;

  /**
   * Maximum number of events kept in the in-tab stream buffer. Oldest events
   * are dropped once the cap is reached.
   */
  maxBufferedEvents: number;
};

export type SseEventRecord = {
  id: string;
  receivedAt: number;
  eventId?: string;
  eventType?: string;
  data: string;
  retry?: number;
};

/**
 * Runtime SSE stream state for the active (or last) connection attempt.
 */
export type SseStreamState = {
  status: SseConnectionStatus;
  connectedAt?: number;
  closedAt?: number;
  statusCode?: number;
  statusText?: string;
  responseHeaders: Array<{ key: string; value: string }>;
  events: SseEventRecord[];
  error?: string;
  /**
   * Count of events accepted after applying `eventTypeFilter`.
   */
  eventCount: number;
  /**
   * Approximate bytes received in event `data` fields.
   */
  bytesReceived: number;
};

export const DEFAULT_SSE_CONFIG: SseConnectionConfig = {
  lastEventId: "",
  eventTypeFilter: "",
  withCredentials: false,
  autoReconnect: true,
  maxBufferedEvents: 500,
};

export const DEFAULT_SSE_STREAM: SseStreamState = {
  status: "idle",
  responseHeaders: [],
  events: [],
  eventCount: 0,
  bytesReceived: 0,
};

File: backend/api/sse.go
package api

// SseConnectPayload starts a long-lived SSE stream identified by ConnectionID
// (typically the workspace tab id).
type SseConnectPayload struct {
	ConnectionID    string            `json:"connectionId"`
	TabID           string            `json:"tabId"`
	URL             string            `json:"url"`
	Headers         map[string]string `json:"headers"`
	LastEventID     string            `json:"lastEventId,omitempty"`
	WithCredentials bool              `json:"withCredentials"`
}

// SseConnectResult is returned immediately when a connect attempt is accepted.
// Handshake details and streamed events follow on Wails runtime events.
type SseConnectResult struct {
	ConnectionID string `json:"connectionId"`
	Error        string `json:"error,omitempty"`
}

type SseHeaderPayload struct {
	Key   string `json:"key"`
	Value string `json:"value"`
}

// SseOpenEventPayload is emitted on event "sse:open" after the HTTP handshake.
type SseOpenEventPayload struct {
	ConnectionID string             `json:"connectionId"`
	TabID        string             `json:"tabId"`
	StatusCode   int                `json:"statusCode"`
	StatusText   string             `json:"statusText"`
	Headers      []SseHeaderPayload `json:"headers"`
	Error        string             `json:"error,omitempty"`
}

// SseStreamEventPayload is emitted on event "sse:event" for each parsed SSE message.
type SseStreamEventPayload struct {
	ConnectionID string `json:"connectionId"`
	TabID        string `json:"tabId"`
	EventID      string `json:"eventId,omitempty"`
	EventType    string `json:"eventType,omitempty"`
	Data         string `json:"data"`
	Retry        *int64 `json:"retry,omitempty"`
}

// SseErrorEventPayload is emitted on event "sse:error".
type SseErrorEventPayload struct {
	ConnectionID string `json:"connectionId"`
	TabID        string `json:"tabId"`
	Error        string `json:"error"`
}

// SseCloseEventPayload is emitted on event "sse:close".
type SseCloseEventPayload struct {
	ConnectionID string `json:"connectionId"`
	TabID        string `json:"tabId"`
}
